src/translator/html.cpp

```diff
@@ -580,6 +589,54 @@ void HTML::copyTaint(Response const &response, std::vector<std::vector<size_t>>
   targetTokenSpans.push_back(sourceTokenSpans[offset]);  // token_tag for ending whitespace
 }
 
+void HTML::annotateTaint(Response const &response, std::vector<SpanIterator> const &targetTokenSpans,
+                         std::vector<HTML::Taint> &targetTokenTags) {
+  auto spanIt = targetTokenSpans.begin();
+  for (size_t sentenceIdx = 0; sentenceIdx < response.target.numSentences(); ++sentenceIdx) {
+    // Sentence prefix
+    targetTokenTags.push_back((*spanIt)->tags);
+    spanIt++;
+
+    // Offset in targetTokenTags at which this sentence's tags start.
+    size_t tagOffset = targetTokenTags.size();
+
+    // Initially, just copy the span's tags to this token
+    for (size_t t = 0; t < response.target.numWords(sentenceIdx); ++t) {
+      targetTokenTags.emplace_back((*spanIt)->tags);
+      spanIt++;
+    }
+
+    // If we have quality score information, add that as metadata as well.
+    if (!response.qualityScores.empty()) {
+      auto const &sentenceQuality = response.qualityScores[sentenceIdx];
+      // Create a single <font> tag for this sentence with sentence level info
+      Tag *sentenceTag = makeTag({Tag::ELEMENT, "font"});
+      sentenceTag->attributes += format(" x-bergamot-sentence-score=\"{}\"", sentenceQuality.sentenceScore);
+
+      // Add that tag to all tokens in this sentence.
+      for (size_t tokenIdx = 0; tokenIdx < response.target.numWords(sentenceIdx); ++tokenIdx) {
+        targetTokenTags[tagOffset + tokenIdx].push_back(sentenceTag);
+      }
+
+      // Add word level <font> tags as well to all tokens that make up a word.
+      for (size_t wordIdx = 0; wordIdx < sentenceQuality.wordRanges.size(); ++wordIdx) {
+        Tag *wordTag = makeTag({Tag::ELEMENT, "font"});
+        wordTag->attributes += format(" x-bergamot-word-score=\"{}\"", sentenceQuality.wordScores[wordIdx]);
+        auto const &range = sentenceQuality.wordRanges[wordIdx];
+        for (size_t tokenIdx = range.begin; tokenIdx < range.end; ++tokenIdx) {
+          targetTokenTags[tagOffset + tokenIdx].push_back(wordTag);
+        }
+      }
+    }
+  }
```

Comment:
What are wordRanges? There are two ways I see this happening for an example of two subwords constituting a word.

(1) `<sw>A</sw></tag><sw>a</sw> -> <w score='s'>A</tag>a</w>` (this is troublesome and can break HTML)
(2)  `<sw>A</sw></tag><sw>a</sw> -> <w score='s'>A</w></tag><w score='s'>a</w>` (this is acceptable if the same value is used across both and rendering (if color coding) appears similar. 

Given `wordIdx` appears to index `sentenceQuality.wordScores` is (1) a possibility?

Reply:
No, (1) not possible thanks to how HTML insertion works. It will automatically generate (2).

There is an additional guarantee right now: the added tag will always be the leaf node. So instead of

(A) `<p><w>He<tag>le</tag>boel</w></p>`  you will have
(B) `<p><w>He</w><tag><w>le</w></tag><w>boel</w></p>` (where those `<w>` tags will have the same attributes again).

This is quite useful since now it wont break CSS selectors like `p > tag { text-decoration: underline; }`.

Reply:
> not possible thanks to how HTML insertion works. It will automatically generate (2).

This is helpful. Thanks. If we are controlling HTML we will need inputs from Mozilla on what their UI plans are, right now we're guessing. Let's carry over this part in #355.